## Bài 1. Tạo Tensor cơ bản
Tạo một Tensor số nguyên kích thước $2 \times 3$ và hiển thị kết quả.
Yêu cầu: Cho biết kích thước, số chiều và kiểu dữ liệu của Tensor a.

In [2]:
import torch

a = torch.tensor(
    [[1, 2, 3],
     [4, 5, 6]],
    dtype=torch.int32
)

print("Tensor a:\n", a)
print("-----------------")

# Trả lời yêu cầu:
print("Kích thước (shape):", a.shape)
print("Số chiều (ndim):", a.ndim)
print("Kiểu dữ liệu (dtype):", a.dtype)


Tensor a:
 tensor([[1, 2, 3],
        [4, 5, 6]], dtype=torch.int32)
-----------------
Kích thước (shape): torch.Size([2, 3])
Số chiều (ndim): 2
Kiểu dữ liệu (dtype): torch.int32


## Bài 2. Tạo các Tensor thường dùng 
Thực hành tạo Tensor ngẫu nhiên, Tensor toàn số 0, toàn số 1 và ma trận đơn vị. 

In [4]:
import torch

# 1. Khởi tạo các ma trận đặc biệt cơ bản (ví dụ với kích thước 3x3)
zeros_tensor = torch.zeros(3, 3)
print("1. Ma trận toàn số 0:\n", zeros_tensor)

ones_tensor = torch.ones(3, 3)
print("\n2. Ma trận toàn số 1:\n", ones_tensor)

eye_tensor = torch.eye(3)
print("\n3. Ma trận đơn vị (đường chéo bằng 1):\n", eye_tensor)

# 2. Khởi tạo ma trận ngẫu nhiên
# Phân bố đều (các giá trị nằm trong khoảng từ 0 đến 1)
rand_tensor = torch.rand(3, 3)
print("\n4. Ma trận ngẫu nhiên đều (torch.rand):\n", rand_tensor)

# Phân bố chuẩn (xoay quanh giá trị 0, có cả số âm và dương)
randn_tensor = torch.randn(3, 3)
print("\n5. Ma trận ngẫu nhiên chuẩn (torch.randn):\n", randn_tensor)

# 3. Code yêu cầu: Tạo ma trận 5x5 toàn số 10
tensor_10_v1 = torch.ones(5, 5) * 10
print("\n6. Ma trận 5x5 toàn số 10 (Cách dùng nhân):\n", tensor_10_v1)


1. Ma trận toàn số 0:
 tensor([[0., 0., 0.],
        [0., 0., 0.],
        [0., 0., 0.]])

2. Ma trận toàn số 1:
 tensor([[1., 1., 1.],
        [1., 1., 1.],
        [1., 1., 1.]])

3. Ma trận đơn vị (đường chéo bằng 1):
 tensor([[1., 0., 0.],
        [0., 1., 0.],
        [0., 0., 1.]])

4. Ma trận ngẫu nhiên đều (torch.rand):
 tensor([[0.6221, 0.4579, 0.1685],
        [0.7925, 0.1133, 0.7739],
        [0.0213, 0.5377, 0.2749]])

5. Ma trận ngẫu nhiên chuẩn (torch.randn):
 tensor([[-0.1265, -2.0247, -1.4892],
        [ 1.1900, -0.5516,  2.4484],
        [ 0.1280,  1.0358, -0.3894]])

6. Ma trận 5x5 toàn số 10 (Cách dùng nhân):
 tensor([[10., 10., 10., 10., 10.],
        [10., 10., 10., 10., 10.],
        [10., 10., 10., 10., 10.],
        [10., 10., 10., 10., 10.],
        [10., 10., 10., 10., 10.]])


- torch.rand(): Sinh ra các giá trị ngẫu nhiên theo phân bố đều (uniform distribution), các số luôn nằm trong khoảng từ 0 đến 1.
- torch.randn(): Sinh ra các giá trị ngẫu nhiên theo phân bố chuẩn (normal/Gaussian distribution), có giá trị trung bình bằng 0 và độ lệch chuẩn bằng 1 (nên kết quả có thể chứa cả số âm và số dương).

## Bài 3. Kiểm tra thông tin Tensor 
Sử dụng shape, size(), ndim và dtype để kiểm tra các thuộc tính của Tensor. 

In [5]:
import torch

a = torch.zeros(2, 3, 4)
print("Tensor a:")
print(a)
print("\n--- Kiểm tra thông tin ---")
print("Shape:", a.shape)
print("Size:", a.size())
print("Number of dimensions (ndim):", a.ndim)
print("Data type:", a.dtype)

Tensor a:
tensor([[[0., 0., 0., 0.],
         [0., 0., 0., 0.],
         [0., 0., 0., 0.]],

        [[0., 0., 0., 0.],
         [0., 0., 0., 0.],
         [0., 0., 0., 0.]]])

--- Kiểm tra thông tin ---
Shape: torch.Size([2, 3, 4])
Size: torch.Size([2, 3, 4])
Number of dimensions (ndim): 3
Data type: torch.float32



Giải thích ý nghĩa của torch.Size([2, 3, 4]):
- Đây là kích thước (shape) của một Tensor 3 chiều (3D Tensor).
- Số 2 đầu tiên đại diện cho số lượng "trang" hoặc "khối" (depth/channels) của Tensor. Nghĩa là Tensor này gồm 2 ma trận con bên trong.
- Số 3 đại diện cho số hàng (rows) của mỗi ma trận con.
- Số 4 đại diện cho số cột (columns) của mỗi ma trận con.
- Tóm lại, Tensor này có chứa 2 khối, mỗi khối là một ma trận kích thước 3x4. Tổng cộng nó chứa 2 * 3 * 4 = 24 phần tử.


## Bài 4. Slicing và Indexing 
Truy xuất một phần dữ liệu từ Tensor nhiều chiều. 

In [36]:
import torch 
 
a = torch.randn(3, 4, 5) 
 
print("Original tensor:") 
print(a) 
 
print("\nTensor a[1]:") 
print(a[1]) 
 
print("\nTensor a[1:, 2:4]:") 
print(a[1:, 2:4]) 

print("1. Shape của a[1]:", a[1].shape) 
print("   Shape của a[1:, 2:4]:", a[1:, 2:4].shape) 
hai_phan_tu_cuoi = a[:, :, -2:]
print("\n2. Lấy hai phần tử cuối cùng trên chiều thứ ba:")
print("- Kích thước (shape) thu được:", hai_phan_tu_cuoi.shape) 
print("- Kết quả chi tiết:\n", hai_phan_tu_cuoi)

Original tensor:
tensor([[[-0.2413, -0.0972,  1.7583, -0.5384,  1.1255],
         [ 2.1858,  0.0639,  0.3233, -0.3329,  0.3710],
         [-1.5217, -0.0166, -0.9552,  1.6311,  1.5069],
         [-1.7504, -0.6219, -1.3503, -0.4080, -0.8320]],

        [[ 1.2199,  0.5983,  2.1219,  0.7601, -0.2133],
         [-0.4809, -0.1463,  1.5642,  1.3596,  0.5052],
         [-1.0602,  0.3013, -0.0803,  0.1061,  1.5667],
         [ 1.5100,  0.0339,  1.5497, -1.4706,  1.0551]],

        [[ 1.7420,  0.4739,  0.6182, -1.6986, -0.9240],
         [-0.8233,  0.8360,  0.5335, -0.5749, -1.1409],
         [ 0.8967, -0.8586, -1.6156, -0.9370,  0.6565],
         [ 0.2995, -0.7581,  0.9999, -0.6850, -0.2552]]])

Tensor a[1]:
tensor([[ 1.2199,  0.5983,  2.1219,  0.7601, -0.2133],
        [-0.4809, -0.1463,  1.5642,  1.3596,  0.5052],
        [-1.0602,  0.3013, -0.0803,  0.1061,  1.5667],
        [ 1.5100,  0.0339,  1.5497, -1.4706,  1.0551]])

Tensor a[1:, 2:4]:
tensor([[[-1.0602,  0.3013, -0.0803,  0.1061,  1.5

## Bài 5. Reshape và Flatten 
Thay đổi hình dạng Tensor mà không làm thay đổi số phần tử. 

In [13]:
import torch
a = torch.randn(3, 4)
so_phan_tu_ban_dau = a.numel() 
a_reshaped = a.reshape(3, 2, 2)
so_phan_tu_luc_sau = a_reshaped.numel()
print("1. Kiểm tra số phần tử:")
print(f"- Số phần tử ban đầu (shape {a.shape}):", so_phan_tu_ban_dau)
print(f"- Số phần tử sau khi reshape (shape {a_reshaped.shape}):", so_phan_tu_luc_sau)

a_2x6 = a.reshape(2, 6)
print("\n2. Thử reshape Tensor thành kích thước 2 x 6:")
print("- Shape thu được:", a_2x6.shape)
print("- Tensor mới hiển thị như sau:\n", a_2x6)

1. Kiểm tra số phần tử:
- Số phần tử ban đầu (shape torch.Size([3, 4])): 12
- Số phần tử sau khi reshape (shape torch.Size([3, 2, 2])): 12

2. Thử reshape Tensor thành kích thước 2 x 6:
- Shape thu được: torch.Size([2, 6])
- Tensor mới hiển thị như sau:
 tensor([[-1.2003,  1.3356,  1.3552,  0.2528,  0.2749, -0.8268],
        [ 1.2794,  1.4386, -0.0592,  1.0702, -1.7362,  0.7858]])


## Bài 6. Phép toán và nhân ma trận 
Thực hiện các phép toán phần tử và phép nhân ma trận bằng toán tử @. 

In [19]:
import torch

a = torch.randn(2, 3)
b = torch.randn(2, 3)
print("matrix A:")
print(a)
print("matrix B:")
print(b)

# Tính toán thử để xem hình thù kết quả
tich_tung_phan_tu = a * b
tich_ma_tran = a @ b.T
print("tich tung phan:") 
print(tich_tung_phan_tu)
print("tich ma tran:")
print(tich_ma_tran)

print("shape của a@b.T: ", tich_ma_tran.shape)

matrix A:
tensor([[ 0.5668,  0.8008, -0.3462],
        [-0.1771,  0.7257,  0.1323]])
matrix B:
tensor([[-0.6790,  0.0511, -0.8057],
        [ 0.0620,  0.4512,  0.2125]])
tich tung phan:
tensor([[-0.3849,  0.0409,  0.2789],
        [-0.0110,  0.3274,  0.0281]])
tich ma tran:
tensor([[-0.0650,  0.3229],
        [ 0.0507,  0.3445]])
shape của a@b.T:  torch.Size([2, 2])


- Phép `a * b` (Nhân element-wise): Là lấy từng phần tử ở cùng vị trí của 2 ma trận nhân trực tiếp với nhau. Yêu cầu hai ma trận (a và b) phải có kích thước y hệt nhau (ở đây cùng là 2x3). Kết quả trả ra một ma trận vẫn giữ nguyên kích thước 2x3.
- Phép `a @ b.T` (Nhân ma trận - Matrix multiplication): Là phép nhân đại số tuyến tính (nhân hàng của ma trận này với cột của ma trận kia). 
  Vì luật nhân ma trận yêu cầu số cột của ma trận đầu phải bằng số hàng của ma trận sau, nên ta không thể nhân a (2x3) với b (2x3) được. Do đó, phải dùng `.T` (Transpose) để đảo ma trận b thành kích thước 3x2. Lúc này (2x3) nhân (3x2) mới hợp lệ.

## Bài 7. Các hàm thống kê trên Tensor 
Tính trung bình, độ lệch chuẩn và tổng tích lũy theo từng cột.

In [21]:
import torch 
 
a = torch.randn(3, 4) 

print(a) 
print("\nMean:") 
print(torch.mean(a, dim=0)) 
 
print("\nStandard deviation:") 
print(torch.std(a, dim=0)) 
 
print("\nCumulative sum:") 
print(torch.cumsum(a, dim=0)) 

tensor([[ 1.0859, -0.9246, -0.5791, -2.8559],
        [-0.0555, -1.1710,  1.1805,  0.5085],
        [ 0.0769, -1.6628, -0.8189,  1.7786]])

Mean:
tensor([ 0.3691, -1.2528, -0.0725, -0.1896])

Standard deviation:
tensor([0.6243, 0.3759, 1.0918, 2.3948])

Cumulative sum:
tensor([[ 1.0859, -0.9246, -0.5791, -2.8559],
        [ 1.0304, -2.0956,  0.6014, -2.3474],
        [ 1.1073, -3.7584, -0.2175, -0.5688]])


- Khi dùng `dim=0`: Phép tính được thực hiện dọc theo các CỘT. Ma trận có 4 cột nên nó sẽ trả về 4 con số (mỗi số là trung bình của một cột). Nó tính toán bằng cách "ép" các hàng lại với nhau.
- Khi dùng `dim=1`: Phép tính được thực hiện ngang qua các HÀNG. Ma trận có 3 hàng nên kết quả sẽ trả về 3 con số (mỗi số là trung bình của một hàng). Nó tính toán bằng cách "ép" các cột lại với nhau.

## Bài 8. Autograd cơ bản 
Tạo biến có requires_grad=True và dùng backward() để tính đạo hàm tự động. 

In [25]:
import torch 
 
x = torch.tensor( 
    3.6, 
    requires_grad=True 
) 
 
y = x * x 
 
y.backward() 
 
print("x =", x) 
print("y =", y) 
print("x.grad =", x.grad) 

x_new = torch.tensor(3.6, requires_grad=True)
y_new = 3 * x_new**2 + 2 * x_new + 1
y_new.backward()
print("x =", x_new) 
print("y =", y_new) 
print("- Kết quả đạo hàm tự động (x_new.grad):", x_new.grad.item())


x = tensor(3.6000, requires_grad=True)
y = tensor(12.9600, grad_fn=<MulBackward0>)
x.grad = tensor(7.2000)
x = tensor(3.6000, requires_grad=True)
y = tensor(47.0800, grad_fn=<AddBackward0>)
- Kết quả đạo hàm tự động (x_new.grad): 23.599998474121094


- Tính bằng tay theo Toán Giải tích: y = x^2  => Đạo hàm y' = 2*x.
- Thay x = 3.6 vào, ta có kết quả: y' = 2 * 3.6 = 7.2.
- So sánh: Kết quả tính tay (7.2) trùng khớp hoàn toàn với kết quả x.grad (7.2000) mà PyTorch tự động tính toán từ hàm y.backward().

## Bài 9. Gradient Descent tìm hệ số đa thức 
Sinh dữ liệu từ đa thức y = x² + 2x + 3, sau đó dùng NAdam và autograd để tìm lại ba hệ số.

In [37]:
import numpy as np
import torch

polynomial = np.poly1d([1, 2, 3])
N = 20
X = np.random.randn(N, 1) * 5
Y = polynomial(X)
XX = np.hstack([
    X * X,
    X,
    np.ones_like(X)
])

w = torch.randn(3, 1, requires_grad=True)
x = torch.tensor(XX, dtype=torch.float32)
y = torch.tensor(Y, dtype=torch.float32)


optimizer = torch.optim.NAdam([w], lr=0.01)

print("--- QUÁ TRÌNH HỌC CỦA MÁY TÍNH ---")
print("1. Ba hệ số đoán mò ban đầu (ngẫu nhiên):")
print(w.detach().numpy())

for epoch in range(1000):
    optimizer.zero_grad()
    y_pred = x @ w
    mse = torch.mean(torch.square(y - y_pred)) # Tính sai số MSE
    mse.backward()
    optimizer.step()

print("\n2. Ba hệ số mà máy tính TỰ TÌM ĐƯỢC sau 1000 vòng học:")
print(w.detach().numpy())


--- QUÁ TRÌNH HỌC CỦA MÁY TÍNH ---
1. Ba hệ số đoán mò ban đầu (ngẫu nhiên):
[[-0.03132661]
 [ 1.4300054 ]
 [-1.0294079 ]]

2. Ba hệ số mà máy tính TỰ TÌM ĐƯỢC sau 1000 vòng học:
[[1.0269836]
 [2.098618 ]
 [0.9540968]]


Nhận xét: Ba hệ số mà AI học được xấp xỉ rất sát với đáp án [1.0, 2.0, 3.0] của hàm đa thức gốc ban đầu. Điều này chứng tỏ thuật toán Gradient Descent đã hoạt động đúng, giúp máy tính từ việc đoán mò ngẫu nhiên ban đầu tiến dần đến được phương trình chính xác nhờ tối ưu hóa sai số.

Giải thích:
1. optimizer.zero_grad(): 
   - Vai trò: Dọn sạch (reset về 0) túi chứa đạo hàm (gradient) của vòng lặp trước đó. 
   - Lý do: PyTorch có cơ chế mặc định là "cộng dồn" đạo hàm. Nếu không dọn sạch đi thì đạo hàm của vòng mới sẽ cộng dính với vòng cũ gây sai số trầm trọng.
   
2. mse.backward(): 
   - Vai trò: Kích hoạt chế độ tính đạo hàm tự động (Autograd). 
   - Lý do: Nó sẽ tính đạo hàm của hàm sai số (mse) theo biến `w`. Ý nghĩa thực tế là nó tìm ra hướng đi (la bàn) để giúp giảm cái sai số này xuống.
   
3. optimizer.step(): 
   - Vai trò: Bước đi cập nhật trọng số. 
   - Lý do: Dựa vào hướng đi đã tính được ở hàm backward, thuật toán optimizer (NAdam) sẽ tiến hành điều chỉnh biến `w` một chút xíu. Cứ lặp lại nhiều lần như vậy, `w` sẽ ngày càng xấp xỉ đáp án đúng.

## Bài 10. Giải hệ phương trình bằng Autograd 
Tối thiểu hóa tổng sai số bình phương để tìm A, B, C, D thỏa bốn phương trình. 

In [30]:
import random
import torch
print("--- TIẾN HÀNH GIẢI HỆ PHƯƠNG TRÌNH ---")
# 1. Khởi tạo 4 biến A, B, C, D ngẫu nhiên (chính là 4 ẩn số cần đi tìm)
A = torch.tensor(random.random(), requires_grad=True)
B = torch.tensor(random.random(), requires_grad=True)
C = torch.tensor(random.random(), requires_grad=True)
D = torch.tensor(random.random(), requires_grad=True)
EPOCHS = 2000
optimizer = torch.optim.NAdam([A, B, C, D], lr=0.01)
# 2. Cho máy tính học (thử nghiệm và sửa sai)
for _ in range(EPOCHS):
    # Đặt 4 phương trình bằng 0 (chuyển vế đổi dấu)
    y1 = A + B - 9
    y2 = C - D - 1
    y3 = A + C - 8
    y4 = B - D - 2
    
    # Sai số bình phương 
    sqerr = y1*y1 + y2*y2 + y3*y3 + y4*y4
    
    optimizer.zero_grad()
    sqerr.backward()      
    optimizer.step()     

A_val = round(A.item(), 4)
B_val = round(B.item(), 4)
C_val = round(C.item(), 4)
D_val = round(D.item(), 4)
print("Kết quả nghiệm tìm được:")
print(f"A = {A_val}, B = {B_val}, C = {C_val}, D = {D_val}")
print("\n" + "="*40)
print("--- TRẢ LỜI YÊU CẦU BÀI 10 ---")
print("="*40)

print("1. Kiểm tra lại nghiệm (thử lại vào phương trình):")
print(f"- Phương trình 1 (A + B = 9): {A_val} + {B_val} = {round(A_val + B_val, 4)}")
print(f"- Phương trình 2 (C - D = 1): {C_val} - {D_val} = {round(C_val - D_val, 4)}")
print(f"- Phương trình 3 (A + C = 8): {A_val} + {C_val} = {round(A_val + C_val, 4)}")
print(f"- Phương trình 4 (B - D = 2): {B_val} - {D_val} = {round(B_val - D_val, 4)}")
print("=> Nhận xét: Máy tính đã tìm ra được bộ nghiệm thỏa mãn 100% cả 4 phương trình trên.")

--- TIẾN HÀNH GIẢI HỆ PHƯƠNG TRÌNH ---
Kết quả nghiệm tìm được:
A = 4.4029, B = 4.597, C = 3.5971, D = 2.597

--- TRẢ LỜI YÊU CẦU BÀI 10 ---
1. Kiểm tra lại nghiệm (thử lại vào phương trình):
- Phương trình 1 (A + B = 9): 4.4029 + 4.597 = 8.9999
- Phương trình 2 (C - D = 1): 3.5971 - 2.597 = 1.0001
- Phương trình 3 (A + C = 8): 4.4029 + 3.5971 = 8.0
- Phương trình 4 (B - D = 2): 4.597 - 2.597 = 2.0
=> Nhận xét: Máy tính đã tìm ra được bộ nghiệm thỏa mãn 100% cả 4 phương trình trên.


Theo Toán Đại số, hệ phương trình này tuy có 4 phương trình nhưng thực chất phương trình số (4) bị phụ thuộc tuyến tính vào 3 phương trình đầu. 
Cụ thể, ta lấy (PT1) - (PT3) + (PT2): 
(A + B) - (A + C) + (C - D) = 9 - 8 + 1 
=> B - D = 2 (Đúng y hệt phương trình số 4).
Vậy hệ thống thực chất chỉ có 3 phương trình độc lập, nhưng lại có tới 4 biến số (A, B, C, D). Quy tắc Toán học chỉ ra rằng: "Hệ có số ẩn nhiều hơn số phương trình độc lập thì hệ đó có VÔ SỐ NGHIỆM".
Vì hệ có vô số nghiệm, nên mỗi lần bạn chạy lại đoạn code trên (với số khởi tạo random ban đầu khác nhau), AI sẽ "dò" ra một bộ nghiệm A,B,C,D hoàn toàn mới nhưng vẫn luôn luôn đúng!

## Bài 11. Viết chương trình PyTorch để thực hiện các yêu cầu sau: 
• Tạo Tensor X có 24 phần tử và reshape lần lượt thành 4 × 6, 2 × 12 và 2 × 3 × 4. 

• Tạo hai Tensor A, B kích thước 3 × 4; tính A+B, A*B, A/B, trung bình theo cột và độ lệch chuẩn theo cột. 

• Cho y = 2x² + 5x + 3. Dùng autograd tính dy/dx tại x = 2 và kiểm tra bằng phép tính thủ công. 

• Thay đa thức trong Bài 9 bằng y = 2x² - 3x + 5; huấn luyện và cho biết các hệ số tìm được. 

In [31]:
import torch
import numpy as np

In [32]:
print("\n--- Ý 1: TẠO TENSOR VÀ RESHAPE ---")
# Tạo mảng chứa 24 con số từ 0 đến 23
X = torch.arange(24) 
print("Tensor gốc (24 phần tử):\n", X)
# Reshape
X_4x6 = X.reshape(4, 6)
print("\n- Reshape thành 4x6:\n", X_4x6)
X_2x12 = X.reshape(2, 12)
print("\n- Reshape thành 2x12:\n", X_2x12)
X_2x3x4 = X.reshape(2, 3, 4)
print("\n- Reshape thành 2x3x4:\n", X_2x3x4)


--- Ý 1: TẠO TENSOR VÀ RESHAPE ---
Tensor gốc (24 phần tử):
 tensor([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17,
        18, 19, 20, 21, 22, 23])

- Reshape thành 4x6:
 tensor([[ 0,  1,  2,  3,  4,  5],
        [ 6,  7,  8,  9, 10, 11],
        [12, 13, 14, 15, 16, 17],
        [18, 19, 20, 21, 22, 23]])

- Reshape thành 2x12:
 tensor([[ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11],
        [12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23]])

- Reshape thành 2x3x4:
 tensor([[[ 0,  1,  2,  3],
         [ 4,  5,  6,  7],
         [ 8,  9, 10, 11]],

        [[12, 13, 14, 15],
         [16, 17, 18, 19],
         [20, 21, 22, 23]]])


In [33]:
print("\n\n--- Ý 2: CÁC PHÉP TOÁN TRÊN TENSOR ---")
A = torch.randn(3, 4)
B = torch.randn(3, 4)
print("Tensor A:\n", A)
print("Tensor B:\n", B)
print("\n- Phép cộng A + B:\n", A + B)
print("- Phép nhân từng phần tử A * B:\n", A * B)
print("- Phép chia từng phần tử A / B:\n", A / B)
# Tính trung bình và độ lệch chuẩn theo cột (dim=0) cho Tensor A
print("\n- Trung bình theo cột của A (Mean dim=0):\n", torch.mean(A, dim=0))
print("- Độ lệch chuẩn theo cột của A (Std dim=0):\n", torch.std(A, dim=0))



--- Ý 2: CÁC PHÉP TOÁN TRÊN TENSOR ---
Tensor A:
 tensor([[-0.4347, -0.8312, -0.2285, -1.0009],
        [ 1.2057, -1.3094,  1.8374,  0.0663],
        [-0.0189,  0.8240, -1.5668, -0.4736]])
Tensor B:
 tensor([[-0.3829,  0.0671, -0.0951,  0.1218],
        [ 1.5908, -1.8212,  0.6683, -0.9924],
        [ 0.8066,  0.5319, -1.0158, -0.8637]])

- Phép cộng A + B:
 tensor([[-0.8175, -0.7641, -0.3235, -0.8791],
        [ 2.7964, -3.1306,  2.5056, -0.9260],
        [ 0.7877,  1.3559, -2.5826, -1.3373]])
- Phép nhân từng phần tử A * B:
 tensor([[ 0.1664, -0.0557,  0.0217, -0.1219],
        [ 1.9180,  2.3846,  1.2278, -0.0658],
        [-0.0152,  0.4383,  1.5915,  0.4091]])
- Phép chia từng phần tử A / B:
 tensor([[  1.1352, -12.3935,   2.4034,  -8.2179],
        [  0.7579,   0.7189,   2.7495,  -0.0668],
        [ -0.0234,   1.5491,   1.5425,   0.5484]])

- Trung bình theo cột của A (Mean dim=0):
 tensor([ 0.2507, -0.4388,  0.0140, -0.4694])
- Độ lệch chuẩn theo cột của A (Std dim=0):
 tensor([0

In [34]:
print("\n\n--- Ý 3: TÍNH ĐẠO HÀM BẰNG AUTOGRAD ---")
# Đặt giá trị x = 2.0 (nhớ dùng số thập phân để PyTorch hiểu là float)
x_val = torch.tensor(2.0, requires_grad=True)
y_val = 2 * x_val**2 + 5 * x_val + 3
# Tính đạo hàm tự động
y_val.backward()
print(f"1. Đạo hàm tự động (Autograd) tại x = 2 là: {x_val.grad.item()}")
# Kiểm tra thủ công
"""
- Tính bằng tay: y' = 4x + 5
- Thay x = 2 vào ta có: y' = 4(2) + 5 = 13
"""
print(f"2. Kiểm tra thủ công: y' = 4*2 + 5 = 13. (Hoàn toàn trùng khớp!)")



--- Ý 3: TÍNH ĐẠO HÀM BẰNG AUTOGRAD ---
1. Đạo hàm tự động (Autograd) tại x = 2 là: 13.0
2. Kiểm tra thủ công: y' = 4*2 + 5 = 13. (Hoàn toàn trùng khớp!)


In [35]:
print("\n\n--- Ý 4: AI TÌM HỆ SỐ ĐA THỨC ---")
# Phương trình gốc là: y = 2x^2 - 3x + 5 (Đáp án chuẩn là 2, -3, 5)
poly_bai11 = np.poly1d([2, -3, 5]) 
N = 20
X_data = np.random.randn(N, 1) * 5
Y_data = poly_bai11(X_data)
# Chẻ dữ liệu đầu vào thành [x^2, x, 1]
XX_data = np.hstack([
    X_data * X_data,
    X_data,
    np.ones_like(X_data)
])
# AI khởi tạo 3 hệ số ngẫu nhiên
w_bai11 = torch.randn(3, 1, requires_grad=True)
x_tensor = torch.tensor(XX_data, dtype=torch.float32)
y_tensor = torch.tensor(Y_data, dtype=torch.float32)
optimizer11 = torch.optim.NAdam([w_bai11], lr=0.01)
# Vòng lặp học tập
for epoch in range(1500): # Tăng lên 1500 vòng để AI ra kết quả mịn hơn
    optimizer11.zero_grad()
    y_predict = x_tensor @ w_bai11
    loss_mse = torch.mean(torch.square(y_tensor - y_predict))
    loss_mse.backward()
    optimizer11.step()
print("- Phương trình gốc có hệ số chuẩn là: [2, -3, 5]")
print("- Ba hệ số AI đã học được là:\n", w_bai11.detach().numpy())
print("=> Nhận xét: Các hệ số học được đã xấp xỉ vô cùng sát với [2.0, -3.0, 5.0]. Mô hình huấn luyện rất thành công!")



--- Ý 4: AI TÌM HỆ SỐ ĐA THỨC ---
- Phương trình gốc có hệ số chuẩn là: [2, -3, 5]
- Ba hệ số AI đã học được là:
 [[ 2.0001547]
 [-3.0001175]
 [ 4.989446 ]]
=> Nhận xét: Các hệ số học được đã xấp xỉ vô cùng sát với [2.0, -3.0, 5.0]. Mô hình huấn luyện rất thành công!
